### DISCUSSION TITLE
Allow creation of read-only users via custom strategy

### DISCUSSION THREAD

**Discussion by cybrox** [outline/outline#2063] [General]
Hi everyone, first of all, thank your for maintaining Outline! :tada: 

The integration of Passport.js and Read-Only users has made Outline a very nice fit for one of my projects. However, I ran into a small inconsistency when setting up auth via a custom passport strategy.

Is there a reason why the account provisioner (`/server/commands/accountProvisioner.js`) and user creator (`server/commands/userCreator.js`) do not support the `isViewer` flag or was this simply not noticed in the recent PR?

If there is no particular reason why this is absent, I'd be happy to PR the required changes.

**Comment by tommoor** [comment]
Hi – simple answer is they don't have that parameter because there is nowhere that would use it. How are you thinking of hooking it up?

**Reply by cybrox** [reply]
Hi Tom,

I created an auth provider based on the passport.js jwt provider. It redirects the user to our internal auth service, which will sign a JWT including the user's "role" and redirect the user back to our in-house Outline installation.

The auth provider then calls the `accountProvisioner` like this:
```js
const result = await accountProvisioner({
  team: {
    name: "Custom Team",
    domain: providerDomain,
    subdomain: providerDomain,
    avatarUrl: CUSTOM_TEAM_ICON
  },
  user: {
    name: payload.name,
    email: payload.email,
    isAdmin: (payload.role == 'admin'),
    isViewer: (payload.role != 'author')   // <--- this is not currently possible
  },
  authenticationProvider: {
    name: 'custom',
    providerId: providerDomain,
  },
  authentication: {
    providerId: `custom:${payload.id}`,
    accessToken: null,
    refreshToken: null,
    scopes: [],
  }
});
```

The possibility to add custom providers is extremely powerful and allows integration of Outline into nearly any system. This is why I would argue that the account provisioner should support all "permission" flags when creating users.

For now, I monkey-patched the two mentioned files to make this work

<details>
    <summary>Patch</summary>
<pre>
diff --git a/server/commands/accountProvisioner.js b/server/commands/accountProvisioner.js
index e05bd614..4daed14c 100644
--- a/server/commands/accountProvisioner.js
+++ b/server/commands/accountProvisioner.js
@@ -73,7 +73,8 @@ export default async function accountProvisioner({
     const result = await userCreator({
       name: userParams.name,
       email: userParams.email,
-      isAdmin: isNewTeam,
+      isAdmin: userParams.isAdmin || false,
+      isViewer: userParams.isViewer || false,
       avatarUrl: userParams.avatarUrl,
       teamId: team.id,
       ip,
diff --git a/server/commands/userCreator.js b/server/commands/userCreator.js
index 9b6785ee..60b85791 100644
--- a/server/commands/userCreator.js
+++ b/server/commands/userCreator.js
@@ -15,6 +15,7 @@ export default async function userCreator({
   name,
   email,
   isAdmin,
+  isViewer,
   avatarUrl,
   teamId,
   authentication,
@@ -23,6 +24,7 @@ export default async function userCreator({
   name: string,
   email: string,
   isAdmin?: boolean,
+  isViewer?: boolean,
   avatarUrl?: string,
   teamId: string,
   ip: string,
@@ -129,6 +131,7 @@ export default async function userCreator({
         name,
         email,
         isAdmin,
+        isViewer,
         teamId,
         avatarUrl,
         service: null,
</pre>
</details>

**Reply by tommoor** [reply]
If you'd like to submit the relevant patch we'd accept it, it will make your code more maintainable in the long term 💪
